feat(serverless): local testing doc (#1386)
scaleway/docs-content: serverless/functions/reference-content/local-testing.mdx

---
meta:
  title: Functions local testing
  description: This section contains information about Serverless Function local development and testing.
content:
  h1: Functions runtimes configuration
  paragraph: This section contains information about Serverless Function local development and testing
tags: functions serverless local testing
dates:
  validation: 2023-03-06
  posted: 2023-03-06
categories:
  - serverless
---

Local testing allows you to run your code on your dev environment with really short deploy time.

Available frameworks :

- [Go](https://github.com/scaleway/serverless-functions-go)
- [Python](https://github.com/scaleway/serverless-functions-python)

# Local testing

Scaleway Serverless Functions can run outside Scaleway infrastructure, this is useful to setup,

> develop and test functions locally to be able to run it directly and analyse logs.

Reviewer: ```suggestion
develop and test functions locally. By running them directly on their machines, users can debug them, and analyse logs.
```

> Advantage of testing locally is to use your favorite debugger and test tools for your handler

Reviewer: ```suggestion
Each offline testing framework is written in the respective language, and emulates the calls made to functions by the Scaleway infrastructure. 
```

> directly to analyse, debug, monitor and benchmark your function.

Reviewer: ```suggestion
The result is a language-specific executable or script, that users can connect to with their favourite debugger and local tools.
```

## Requirements

Local testing your function will create a local web server that listens to a given port.

You will need to be able to install packages with the dependency manager of the runtime you use 
(ex: `pip`, `go get`...).

> ## Limits

Reviewer: ```suggestion
## Limitations
```

> Building a function locally does not guarantee your function will work on Scaleway Serverless Functions.

Reviewer: ```suggestion
Each offline testing library does its best to replicate the Scaleway Functions environment, but cannot fully emulate the surrounding infrastructure. Therefore, although being able to run your function locally is a good indicator that it will work, it may still fail due to the execution environment used in Scaleway Serverless Functions.
```

> This is because we build your function on our build pipelines and it can differs 

Reviewer: ```suggestion
The most significant difference is that when running in Scaleway Serverless Functions, we have packaged your function into a custom Docker image, built in our build pipeline, 
```

> about operating system and system libraries available on the build image.

Reviewer: ```suggestion
This means that at runtime, your function may be launched in a different operating system, with different libraries available to your local development environment.
```

> Performances can be different as Serverless Functions involves some [Limitations](/serverless/functions/reference-content/functions-limitations/)

Reviewer: ```suggestion
Performance during local testing can differ from the deployed Serverless Functions environment, and will involve [limitations](/serverless/functions/reference-content/functions-limitations/) around resource usage and quotas.
```

tags: functions serverless local testing
dates:
  validation: 2023-03-06
  posted: 2023-03-06
categories:
  - serverless
---

Local testing allows you to run your code on your dev environment with really short deploy time.

Available frameworks:

- [Go](https://github.com/scaleway/serverless-functions-go)
- [Python](https://github.com/scaleway/serverless-functions-python)

# Local testing

Scaleway Serverless Functions can run outside Scaleway infrastructure, this is useful to setup,
develop and test functions locally. By running them directly on their machines, users can debug them, and analyse logs.

Each local testing framework is written in the respective language, and emulates the calls made to functions by the Scaleway infrastructure.
The result is a language-specific executable or script, that users can connect to with their favourite debugger and local tools.

## Requirements

Local testing your function will create a local web server that listens to a given port.

You will need to be able to install packages with the dependency manager of the runtime you use 
(ex: `pip`, `go get`...).

## Limitations

> Each local testing library does its best to replicate the Scaleway Functions environment, but cannot fully emulate 

Reviewer 2: ```suggestion
Each local testing library does its best to replicate the Serverless Functions environment, but cannot fully emulate 
```

the surrounding infrastructure. Therefore, although being able to run your function locally is a good indicator 
that it will work, it may still fail due to the execution environment used in Scaleway Serverless Functions.

The most significant difference is that when running in Scaleway Serverless Functions, 

> we have packaged your function into a custom Docker image, built in our build pipeline, 

Reviewer 2: ```suggestion
we have packaged your function into a custom Docker image, built in our build pipeline. 
```
